src/common/tusb_compiler.h

```diff
@@ -123,6 +136,77 @@
 
   #define TU_BSWAP16(u16) (__iar_builtin_REV16(u16))
   #define TU_BSWAP32(u32) (__iar_builtin_REV(u32))
+
+#elif defined(__CCRX__)
+  #define TU_ATTR_ALIGNED(Bytes)
+  #define TU_ATTR_SECTION(sec_name)
+  #define TU_ATTR_PACKED
+  #define TU_ATTR_WEAK
+  #define TU_ATTR_ALWAYS_INLINE
+  #define TU_ATTR_DEPRECATED(mess)
+  #define TU_ATTR_UNUSED
+  #define TU_ATTR_USED
+
+  #define TU_PACK_STRUCT_BEGIN _Pragma("pack")
+  #define TU_PACK_STRUCT_END _Pragma("packoption")
+
+  #define TU_BIT_FIELD_ORDER_BEGIN _Pragma("bit_order right")
+  #define TU_BIT_FIELD_ORDER_END _Pragma("bit_order")
+
+  // Endian conversion use well-known host to network (big endian) naming
+  #if defined(__LIT)
+    #define TU_BYTE_ORDER TU_LITTLE_ENDIAN
+    #define TU_ENDIAN_LITTLE_BEGIN
+    #define TU_ENDIAN_LITTLE_END
+  #else
+    #define TU_BYTE_ORDER TU_BIG_ENDIAN
+    #define TU_ENDIAN_LITTLE_BEGIN _Pragma("endian little")
+    #define TU_ENDIAN_LITTLE_END _Pragma("endian")
+  #endif
+
+  #define TU_BSWAP16(u16) ((unsigned short)_builtin_revw((unsigned long)u16))
+  #define TU_BSWAP32(u32) (_builtin_revl(u32))
+
+  /* activate the "aligned" emulation, because this toolchain does not know
+     the aligned attribute (or something similar yet) */
+  #define TU_HAS_NO_ATTR_ALIGNED
+
+  /* activate the "weak" function emulation, because this toolchain does not
+     know the weak attribute (or something similar yet) */
+  #define TU_HAS_NO_ATTR_WEAK
+
+  // make sure to define away the standard definition of this macro
+  #undef MAKE_WEAK_FUNC
+  // Helper macro to construct a "weak" function name
+  #define MAKE_WEAK_FUNC(func_name)  weak_ ## func_name
+
+  #if defined(TU_HAS_NO_ATTR_WEAK)
+  // "Weak function" emulation defined in cdc_device.h
+  #define ADD_WEAK_FUNC_TUD_CDC_RX_CB           0
+  #define ADD_WEAK_FUNC_TUD_CDC_RX_WANTED_CB    0
+  #define ADD_WEAK_FUNC_TUD_CDC_TX_COMPLETE_CB  0
+  #define ADD_WEAK_FUNC_TUD_CDC_LINE_STATE_CB   0
+  #define ADD_WEAK_FUNC_TUD_CDC_LINE_CODING_CB  0
+  #define ADD_WEAK_FUNC_TUD_CDC_SEND_BREAK_CB   0
+
+  // "Weak function" emulation defined in usbd_pvt.h
+  #define ADD_WEAK_FUNC_USBD_APP_DRIVER_GET_CB  0
+
+  // "Weak function" emulation defined in usbd.h
+  #define ADD_WEAK_FUNC_TUD_DESCRIPTOR_BOS_CB               0
+  #define ADD_WEAK_FUNC_TUD_DESCRIPTOR_DEVICE_QUALIFIER_CB  0
+  #define ADD_WEAK_FUNC_TUD_MOUNT_CB                        0
+  #define ADD_WEAK_FUNC_TUD_UMOUNT_CB                       0
+  #define ADD_WEAK_FUNC_TUD_SUSPEND_CB                      0
+  #define ADD_WEAK_FUNC_TUD_RESUME_CB                       0
+  #define ADD_WEAK_FUNC_TUD_VENDOR_CONTROL_XFER_CB          0
+
+  // "Weak function" emulation defined in dcd.h
+  #define ADD_WEAK_FUNC_DCD_EDPT0_STATUS_COMPLETE  0
+  #define ADD_WEAK_FUNC_DCD_EDPT_CLOSE             0
+  #define ADD_WEAK_FUNC_DCD_EDPT_XFER_FIFO         0
```

Comment:
This is troublesome, I don't think we could replicate this for all the stack, nor I could keep it consistent after this PR si merge. And explicitly requires user to manually enable each callback is too much. This is deal breaker,

Reply:
Ok, I understand that. The advantage of this would be, that the compiler does not generate code if the callback is defined as NULL.

Reply:
It is not convenience to force user to manually enable this one by one. Also the modification is too much to support and maintain. There are dozens more used by other class driver and host stack in the future. For ccrx, without weak support, I guess this is good compromise.